### ISSUE TITLE
doctest is broken with cabal 3.0 (?)

### ISSUE THREAD

**Issue by googleson78** [sol/doctest#245]
There seems to be some difference in behaviour, perhaps what environment is sought for packages?? I honestly have no idea.

I am asking here in case someone is aware of such a (potential) problem.

Example:
https://travis-ci.org/polysemy-research/polysemy/jobs/600875499

**Comment by travitch** [comment]
For whatever it is worth, I'm also seeing these failures with cabal 3.0 and ghc 8.6.5

**Comment by quasicomputational** [comment]
Yeah, `doctest` itself doesn't know about where various build tools put package databases - it punts to GHC, which in turn checks some environment variables. The best solution for now would be asking upstream to use `cabal-doctest`, which is a hack that does the necessary wiring-up in a build-tool-independent manner; a cabal-install solution is to use `cabal v2-exec` to invoke `doctest`, which sets the appropriate environment variables.

**Comment by hasufell** [comment]
>  a cabal-install solution is to use `cabal v2-exec` to invoke `doctest`, which sets the appropriate environment variables.

`cabal new-exec doctest` doesn't work with `test-suite doctest` after building.

**Comment by travitch** [comment]
A coworker clued me in to my problem: cabal 3.0 no longer creates the GHC environment files by default (cabal 2.0 did).  If I build my project with `--write-ghc-environment-files=always`, my doctests are able to find the packages they import.  I ended up putting this in my ~/.cabal/config to set it globally.

**Comment by simonmichael** [comment]
@travitch, thanks for that. Some more notes:
- you can put `write-ghc-environment-files: always` in your project's `cabal.project` file
- cabal 2 won't understand that, but it's just a warning so shouldn't stop it
- cabal 3 will generate the .ghc.environment* file, but the first run of a doctest suite will still fail; the second run will succeed. Hopefully in practice cabal build will generate the file in time for the test suite.

**Comment by TomMD** [comment]
Using ghc environment files is a good way to get further.  It results in doctest knowing where to find the packages.  We still need doctest to only use the packages of interest for a given build unit or else it has trouble with duplicate modules.

As an example, this problem is observable when you have crypto-api and cryptonite installed then import `Crypto.Random`.
